File: 03-Azure/01-01-App Innovation/05_AppModernizationWithSecurity/dotnet/src/LegoCatalog.App/Configuration/CatalogRuntimeOptions.cs
using Microsoft.Data.SqlClient;

namespace LegoCatalog.App.Configuration;

/// <summary>
/// Holds validated process configuration shared by catalog services.
/// </summary>
public sealed record CatalogRuntimeOptions(
    string SqlConnectionString,
    string ImagesPath,
    string SeedPath,
    bool StartupImportEnabled,
    string PerformanceApiKey,
    int PerformanceWorkFactor,
    string ServiceVersion,
    string DeploymentEnvironment,
    string RevisionName)
{
    public const int DefaultWorkFactor = 10;
    public const int MaximumWorkFactor = 25;

    /// <summary>
    /// Reads and validates the frozen environment/configuration contract.
    /// </summary>
    public static CatalogRuntimeOptions Load(
        IConfiguration configuration,
        string contentRootPath)
    {
        var host = Required(
            configuration,
            "CATALOG_DATABASE_HOST",
            "Catalog:Database:Host");
        var database = Required(
            configuration,
            "CATALOG_DATABASE_NAME",
            "Catalog:Database:Name");
        var username = Optional(
            configuration,
            "CATALOG_DATABASE_USERNAME",
            "Catalog:Database:Username");
        var password = Optional(
            configuration,
            "CATALOG_DATABASE_PASSWORD",
            "Catalog:Database:Password");
        if (string.IsNullOrWhiteSpace(username) != string.IsNullOrWhiteSpace(password))
        {
            throw new InvalidOperationException(
                "CATALOG_DATABASE_USERNAME and CATALOG_DATABASE_PASSWORD must be configured together.");
        }

        var port = ParseOptionalPort(configuration["CATALOG_DATABASE_PORT"]);
        var isManagedHost = host.EndsWith(
            ".database.windows.net",
            StringComparison.OrdinalIgnoreCase);
        var connection = new SqlConnectionStringBuilder
        {
            DataSource = port is null ? host : $"{host},{port}",
            InitialCatalog = database,
            Encrypt = isManagedHost,
            TrustServerCertificate = !isManagedHost,
            ConnectTimeout = 5,
            ApplicationName = "mh-catalog-dotnet",
            MultipleActiveResultSets = false,
            PersistSecurityInfo = false,
        };
        if (username is null)
        {
            connection.IntegratedSecurity = true;
        }
        else
        {
            connection.UserID = username;
            connection.Password = password;
        }

        var imagesPath = ResolvePath(
            Required(configuration, "CATALOG_IMAGES_PATH", "Catalog:ImagesPath"),
            contentRootPath);
        var seedPath = ResolvePath(
            Required(configuration, "CATALOG_SEED_PATH", "Catalog:SeedPath"),
            contentRootPath);
        var startupImportEnabled = ParseBoolean(
            configuration["CATALOG_STARTUP_IMPORT_ENABLED"]
                ?? configuration["Catalog:StartupImportEnabled"],
            defaultValue: true,
            "CATALOG_STARTUP_IMPORT_ENABLED");
        var apiKey = Required(configuration, "PERFTEST_API_KEY");
        var workFactor = ParseWorkFactor(configuration["PERFTEST_WORK_FACTOR"]);
        var deploymentEnvironment = Required(configuration, "DEPLOYMENT_ENVIRONMENT");
        if (!string.Equals(deploymentEnvironment, "lab", StringComparison.Ordinal))
        {
            throw new InvalidOperationException(
                "DEPLOYMENT_ENVIRONMENT must be 'lab' for the workshop contract.");
        }

        return new CatalogRuntimeOptions(
            connection.ConnectionString,
            imagesPath,
            seedPath,
            startupImportEnabled,
            apiKey,
            workFactor,
            Required(configuration, "OTEL_SERVICE_VERSION"),
            deploymentEnvironment,
            Required(configuration, "CONTAINER_APP_REVISION"));
    }

    /// <summary>
    /// Parses the bounded performance work factor.
    /// </summary>
    public static int ParseWorkFactor(string? rawValue)
    {
        if (string.IsNullOrWhiteSpace(rawValue))
        {
            return DefaultWorkFactor;
        }

        if (!int.TryParse(rawValue, out var parsed)
            || parsed < 1
            || parsed > MaximumWorkFactor)
        {
            throw new InvalidOperationException(
                $"PERFTEST_WORK_FACTOR must be an integer from 1 through {MaximumWorkFactor}.");
        }

        return parsed;
    }

    private static int? ParseOptionalPort(string? rawValue)
    {
        if (string.IsNullOrWhiteSpace(rawValue))
        {
            return null;
        }

        if (!int.TryParse(rawValue, out var port) || port is < 1 or > 65535)
        {
            throw new InvalidOperationException(
                "CATALOG_DATABASE_PORT must be an integer from 1 through 65535.");
        }

        return port;
    }

    private static bool ParseBoolean(
        string? rawValue,
        bool defaultValue,
        string name)
    {
        if (string.IsNullOrWhiteSpace(rawValue))
        {
            return defaultValue;
        }

        if (!bool.TryParse(rawValue, out var parsed))
        {
            throw new InvalidOperationException($"{name} must be true or false.");
        }

        return parsed;
    }

    private static string ResolvePath(string path, string contentRootPath)
    {
        return Path.GetFullPath(
            Path.IsPathRooted(path) ? path : Path.Combine(contentRootPath, path));
    }

    private static string Required(
        IConfiguration configuration,
        string key,
        string? fallbackKey = null)
    {
        return Optional(configuration, key, fallbackKey)
            ?? throw new InvalidOperationException($"{key} is required.");
    }

    private static string? Optional(
        IConfiguration configuration,
        string key,
        string? fallbackKey = null)
    {
        var value = configuration[key]
            ?? (fallbackKey is null ? null : configuration[fallbackKey]);
        return string.IsNullOrWhiteSpace(value) ? null : value.Trim();
    }
}

File: 03-Azure/01-01-App Innovation/05_AppModernizationWithSecurity/java/src/main/java/com/microsoft/microhack/catalog/config/CatalogRuntimeOptions.java
package com.microsoft.microhack.catalog.config;

import java.nio.file.Path;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.Objects;
import org.springframework.core.env.Environment;

/** Holds the validated and bounded runtime configuration. */
public record CatalogRuntimeOptions(
        String databaseHost,
        String databaseName,
        Path imagesPath,
        Path seedPath,
        boolean startupImportEnabled,
        String performanceApiKey,
        int performanceWorkFactor,
        String serviceVersion,
        String deploymentEnvironment,
        String revisionName,
        String serviceInstanceId,
        String otlpEndpoint) {

    public static final String SERVICE_NAME = "mh-catalog-java";
    public static final String SERVICE_NAMESPACE = "app-innovation";
    public static final int DEFAULT_WORK_FACTOR = 10;
    public static final int MAXIMUM_WORK_FACTOR = 25;

    /** Reads and validates every application and resource variable. */
    public static CatalogRuntimeOptions from(Environment environment, String instanceId) {
        String databaseHost = require(environment, "CATALOG_DATABASE_HOST");
        parsePort(environment.getProperty("CATALOG_DATABASE_PORT"));
        String databaseName = require(environment, "CATALOG_DATABASE_NAME");
        require(environment, "CATALOG_DATABASE_USERNAME");
        require(environment, "CATALOG_DATABASE_PASSWORD");

        String deploymentEnvironment = require(environment, "DEPLOYMENT_ENVIRONMENT");
        if (!"lab".equals(deploymentEnvironment)) {
            throw new IllegalStateException("DEPLOYMENT_ENVIRONMENT must be 'lab'.");
        }
        String apiKey = require(environment, "PERFTEST_API_KEY");
        if (apiKey.length() > 1024) {
            throw new IllegalStateException("PERFTEST_API_KEY must not exceed 1024 characters.");
        }
        return new CatalogRuntimeOptions(
                databaseHost,
                databaseName,
                Path.of(require(environment, "CATALOG_IMAGES_PATH")).toAbsolutePath().normalize(),
                Path.of(require(environment, "CATALOG_SEED_PATH")).toAbsolutePath().normalize(),
                parseBoolean(
                        environment.getProperty("CATALOG_STARTUP_IMPORT_ENABLED"),
                        "CATALOG_STARTUP_IMPORT_ENABLED"),
                apiKey,
                parseWorkFactor(environment.getProperty("PERFTEST_WORK_FACTOR")),
                bounded(require(environment, "OTEL_SERVICE_VERSION"), "OTEL_SERVICE_VERSION", 128),
                deploymentEnvironment,
                bounded(require(environment, "CONTAINER_APP_REVISION"), "CONTAINER_APP_REVISION", 128),
                bounded(Objects.requireNonNull(instanceId), "service.instance.id", 128),
                parseEndpoint(require(environment, "OTEL_EXPORTER_OTLP_ENDPOINT")));
    }

    /** Parses the bounded performance work factor. */
    public static int parseWorkFactor(String rawValue) {
        if (rawValue == null || rawValue.isBlank()) {
            return DEFAULT_WORK_FACTOR;
        }
        try {
            int value = Integer.parseInt(rawValue);
            if (value < 1 || value > MAXIMUM_WORK_FACTOR) {
                throw new IllegalStateException(workFactorMessage());
            }
            return value;
        } catch (NumberFormatException exception) {
            throw new IllegalStateException(workFactorMessage(), exception);
        }
    }

    private static String workFactorMessage() {
        return "PERFTEST_WORK_FACTOR must be an integer from 1 through "
                + MAXIMUM_WORK_FACTOR + ".";
    }

    private static boolean parseBoolean(String rawValue, String name) {
        if (rawValue == null || rawValue.isBlank()) {
            return true;
        }
        if ("true".equalsIgnoreCase(rawValue)) {
            return true;
        }
        if ("false".equalsIgnoreCase(rawValue)) {
            return false;
        }
        throw new IllegalStateException(name + " must be true or false.");
    }

    private static void parsePort(String rawValue) {
        if (rawValue == null || rawValue.isBlank()) {
            return;
        }
        try {
            int port = Integer.parseInt(rawValue);
            if (port < 1 || port > 65535) {
                throw new IllegalStateException("CATALOG_DATABASE_PORT must be from 1 through 65535.");
            }
        } catch (NumberFormatException exception) {
            throw new IllegalStateException(
                    "CATALOG_DATABASE_PORT must be from 1 through 65535.", exception);
        }
    }

    private static String require(Environment environment, String name) {
        return bounded(environment.getProperty(name), name, 4096);
    }

    private static String bounded(String value, String name, int maximumLength) {
        if (value == null || value.isBlank()) {
            throw new IllegalStateException(name + " is required.");
        }
        String trimmed = value.trim();
        if (trimmed.length() > maximumLength) {
            throw new IllegalStateException(name + " exceeds its maximum length.");
        }
        return trimmed;
    }

    private static String parseEndpoint(String value) {
        try {
            URI uri = new URI(value);
            if (!uri.isAbsolute()
                    || (!"http".equalsIgnoreCase(uri.getScheme())
                    && !"https".equalsIgnoreCase(uri.getScheme()))) {
                throw new IllegalStateException(
                        "OTEL_EXPORTER_OTLP_ENDPOINT must be an absolute HTTP(S) URI.");
            }
            return uri.toString();
        } catch (URISyntaxException exception) {
            throw new IllegalStateException(
                    "OTEL_EXPORTER_OTLP_ENDPOINT must be an absolute HTTP(S) URI.", exception);
        }
    }
}
